### PR TITLE
Add distinguish action to posts

### PR DESCRIPTION
**Pull request by scarow**
comments:
https://reddit.atlassian.net/browse/CE-384
https://reddit.atlassian.net/browse/CE-374

posts:
https://reddit.atlassian.net/browse/CE-368
https://reddit.atlassian.net/browse/CE-379

### CODE REVIEW

**Comment by scarow** [comment]
👓 @birakattack @madbook 

Notes: 
1. The `distinguish icon` is borked --  ignore for now. This should be an easy fix, just want to get the logic review out of the way.
2. You must pull & link to `node-api-client` https://github.com/reddit/node-api-client/pull/188

**Comment by scarow** [comment]
toatssitst_-_toasties_test

File: src/app/actions/modTools.js [hunk 1]
```diff
@@ -101,6 +123,27 @@ export const approve = id => async (dispatch, getState) => {
...
+  const apiOptions = apiOptionsFromState(state);
+  const model = modelFromThingId(id, state);
+
+  dispatch(distinguishPending());
+
+  try {
+    await Modtools.distinguish(apiOptions, id, how);
+    // DISTINGUISH_TYPE[how] will return either 'moderator' or '', which
+    // is added to model.distinguished state
+    dispatch(distinguishSuccess(model, DISTINGUISH_TYPE[how]));
```
Comment (madbook) [inline src/app/actions/modTools.js]: I think this is a little confusing – `distinguish` here has an arg `how` that is expected to be the string `yes` or `no`, while the action creator it calls (`distinguishSuccess`) also has an argument called `how` that expects to either be the string `moderator` or an empty string.  I think we could drop a little complexity by either:

1. replacing the yes/no strings with booleans, changing the argument to this method to something like `isDistinguished` and choosing which string to pass along to `distinguishSuccess` locally, or
2. dropping the yes/no strings and just passing either `moderator` or an empty string directly into this method.

File: src/app/components/Comment/CommentDropdown/index.jsx [hunk 2]
```diff
@@ -93,6 +97,8 @@ CommentDropdown.propTypes = {
   removedBy: T.string,
   showModModal: T.bool,
   modModalId: T.string,
+  isDistinguished: T.string,
```
Comment (madbook) [inline src/app/components/Comment/CommentDropdown/index.jsx]: related to my above comment, it's a bit unexpected for a property name starting with `is` to be something other than a bool.  Maybe something like `distinguishType` for a string? (or keep it as is and change the type to bool)

File: src/app/components/ModeratorModal/index.jsx [hunk 3]
```diff
@@ -9,8 +10,20 @@ import { getStatusBy, getApprovalStatus } from 'lib/modToolHelpers.js';
 import * as modActions from 'app/actions/modTools';
 
 const T = React.PropTypes;
+const DISTINGUISH_FALSEY_VALS = ['', 'no', undefined, null];
 
 export class ModeratorModal extends React.Component {
+
+  onDistinguish(isDistinguished) {
+    const how = includes(DISTINGUISH_FALSEY_VALS, isDistinguished) ? 'yes' : 'no';
```
Comment (madbook) [inline src/app/components/ModeratorModal/index.jsx]: I think some of this can also go away RE above comments.

**Comment by scarow** [comment]
With separation between toggle/non-toggle items:

screen shot 2017-01-27 at 11 15 56 am

**Comment by scarow** [comment]
:haircut_man: see new node-api-client changes as well

File: src/app/components/ModeratorModal/index.jsx [hunk 4]
```diff
@@ -31,24 +49,35 @@ export class ModeratorModal extends React.Component {
...
-                onClick={ this.props.onSpam }
-                isSelected={ this.props.isSpam }
-              />
-              <DropdownRow
-                icon='check-circled'
-                text='Approve'
-                onClick={ this.props.onApprove }
-                isSelected={ this.props.isApproved }
-              />
+              { this.props.isMine
```
Comment (birakattack) [inline src/app/components/ModeratorModal/index.jsx]: Why is this needed if it is only shown in the `ModeratorModal`?

File: src/app/actions/modTools.js [hunk 5]
```diff
@@ -47,6 +50,20 @@ export const approvalSuccess = (thing, username) => ({
   username,
 });
 
+export const distinguishPending = () => ({
+  type: MODTOOLS_DISTINGUISH_PENDING,
+});
+
+export const distinguishError = () => ({
```
Comment (birakattack) [inline src/app/actions/modTools.js]: The action payload for this should be 
```
{
    type: ...,
    error,
    message: 'Some message goes here'
}
```

We can then listen for this action in `app/reducer/toaster.js` to emit a toast that contains the message passed in the above payload.

File: src/app/actions/modTools.js [hunk 6]
```diff
@@ -47,6 +50,20 @@ export const approvalSuccess = (thing, username) => ({
   username,
 });
 
+export const distinguishPending = () => ({
```
Comment (birakattack) [inline src/app/actions/modTools.js]: We can probably just get rid of this, since this is not doing anything.

File: src/app/actions/modTools.js [hunk 7]
```diff
@@ -12,6 +12,9 @@ export const MODTOOLS_REMOVAL_SUCCESS = 'MODTOOLS_REMOVAL_SUCCESS';
 export const MODTOOLS_APPROVAL_PENDING = 'MODTOOLS_APPROVAL_PENDING';
 export const MODTOOLS_APPROVAL_ERROR = 'MODTOOLS_APPROVAL_ERROR';
 export const MODTOOLS_APPROVAL_SUCCESS = 'MODTOOLS_APPROVAL_SUCCESS';
+export const MODTOOLS_DISTINGUISH_PENDING = 'MODTOOLS_DISTINGUISH_PENDING'
```
Comment (birakattack) [inline src/app/actions/modTools.js]: We can remove this since it's not really being used.

File: src/app/actions/modTools.js [hunk 8]
```diff
@@ -101,6 +118,25 @@ export const approve = id => async (dispatch, getState) => {
   }
 };
 
+export const distinguish = (id, distinguishType) => async (dispatch, getState) => {
+  const state = getState();
+  const apiOptions = apiOptionsFromState(state);
+  const model = modelFromThingId(id, state);
+
+  dispatch(distinguishPending());
```
Comment (birakattack) [inline src/app/actions/modTools.js]: We can remove this as well.

File: src/app/components/ModeratorModal/index.jsx [hunk 9]
```diff
@@ -69,12 +98,15 @@ ModeratorModal.propTypes = {
   isSpam: T.bool.isRequired,
   removedBy: T.string,
   approvedBy: T.string,
+  isDistinguished: T.string,
+  isMine: T.bool,
```
Comment (birakattack) [inline src/app/components/ModeratorModal/index.jsx]: We can remove this.

File: src/app/components/Post/PostDropdown/index.jsx [hunk 10]
```diff
@@ -35,6 +35,8 @@ export default function PostDropdown(props) {
     removedBy,
     showModModal,
     modModalId,
+    isDistinguished,
+    isMine,
```
Comment (birakattack) [inline src/app/components/Post/PostDropdown/index.jsx]: Just adding a note to remove.

File: src/app/components/Post/PostDropdown/index.jsx [hunk 11]
```diff
@@ -60,6 +62,8 @@ export default function PostDropdown(props) {
         isSpam={ isSpam }
         approvedBy={ approvedBy }
         removedBy={ removedBy }
+        isMine={ isMine }
```
Comment (birakattack) [inline src/app/components/Post/PostDropdown/index.jsx]: Remove.

File: src/app/components/Post/PostFooter/index.jsx [hunk 12]
```diff
@@ -112,6 +112,8 @@ export default class PostFooter extends React.Component {
         removedBy={ post.bannedBy }
         showModModal={ showModModal }
         modModalId={ modModalId }
+        isDistinguished={ post.distinguished }
+        isMine={ user && user.name === post.author }
```
Comment (birakattack) [inline src/app/components/Post/PostFooter/index.jsx]: Remove.

File: src/app/components/ModeratorModal/index.jsx [hunk 13]
```diff
@@ -10,14 +10,19 @@ import { getStatusBy, getApprovalStatus } from 'lib/modToolHelpers.js';
 import * as modActions from 'app/actions/modTools';
 
 const T = React.PropTypes;
-const DISTINGUISH_FALSEY_VALS = ['', 'no', undefined, null];
+const DISTINGUISH_FALSEY_VALS = ['', undefined, null];
```
Comment (madbook) [inline src/app/components/ModeratorModal/index.jsx]: You shouldn't need to do this anymore – since all of these values are already falsey, you should be able to just do

```javascript
const distinguishType = !isDistinguished ? DISTINGUISH_TYPES.MODERATOR : DISTINGUISH_TYPES.NONE;
```

File: src/app/components/Post/PostHeader/index.jsx [hunk 14]
```diff
@@ -215,10 +215,10 @@ function renderApprovalStatusFlair(post) {
 
   return (
     <span className='PostHeader__approval-status-flair'>
+      { distinguished === 'moderator' ? MOD_FLAIR : null }
```
Comment (madbook) [inline src/app/components/Post/PostHeader/index.jsx]: can this use `DISTINGUISH_TYPES.MODERATOR` instead of the hard-coded string?

File: src/app/actions/modTools.js [hunk 15]
```diff
@@ -63,10 +58,10 @@ export const distinguishError = () => ({
   type: MODTOOLS_DISTINGUISH_ERROR,
 });
 
-export const distinguishSuccess = (thing, how) => ({
+export const distinguishSuccess = (thing, distinguishType) => ({
```
Comment (madbook) [inline src/app/actions/modTools.js]: 👍 👏

File: src/app/components/ModeratorModal/index.jsx (hunk 4 above)
Comment (scarow) [inline src/app/components/ModeratorModal/index.jsx]: `isMine` refers to a post or comment that is yours (not a subreddit). As in, I can only distinguish a post or comment that i made.

File: src/app/components/ModeratorModal/index.jsx (hunk 9 above)
Comment (scarow) [inline src/app/components/ModeratorModal/index.jsx]: Why?

File: src/app/components/Post/PostFooter/index.jsx (hunk 12 above)
Comment (scarow) [inline src/app/components/Post/PostFooter/index.jsx]: Mentioned above -- we need `isMine`

**Comment by scarow** [comment]
:haircut_man: thanks guys!

File: src/app/components/ModeratorModal/index.jsx [hunk 16]
```diff
@@ -9,8 +10,25 @@ import { getStatusBy, getApprovalStatus } from 'lib/modToolHelpers.js';
 import * as modActions from 'app/actions/modTools';
 
 const T = React.PropTypes;
+const DISTINGUISH_FALSEY_VALS = ['', undefined, null];
```
Comment (madbook) [inline src/app/components/ModeratorModal/index.jsx]: Is there a reason to leave this in?

**Comment by madbook** [comment]
whoops, 💅 for harold update and slack ping

File: src/app/components/ModeratorModal/index.jsx (hunk 16 above)
Comment (scarow) [inline src/app/components/ModeratorModal/index.jsx]: ehhh I suppose not. I guess I could just check which value it currently is, then toggle it

**Comment by scarow** [comment]
:haircut_man:

File: src/app/components/ModeratorModal/index.jsx (hunk 4 above)
Comment (birakattack) [inline src/app/components/ModeratorModal/index.jsx]: Ahh got it sorry for the noise on that

Comment (birakattack) [inline src/app/components/ModeratorModal/index.jsx]: Would it be better to move this into a function like `renderDistinguishButton(props)` and pass in the props to render the button. That way in the main render method we would just call `{ renderDistinguishButton(this.props) }` instead of the largish ternary operator?

File: src/app/components/Post/PostHeader/index.jsx [hunk 17]
```diff
@@ -202,16 +201,24 @@ function renderPostFlair(post, single) {
   );
 }
 
-function renderApprovalStatusFlair(isApproved, isRemoved, isSpam) {
-  if (!(isApproved || isRemoved || isSpam)) {
+function renderApprovalStatusFlair(post) {
```
Comment (birakattack) [inline src/app/components/Post/PostHeader/index.jsx]: This should be renamed if it is being overloaded to items other than approve, remove and spam. Maybe something like `renderPostIcons`?

File: src/app/components/Post/PostHeader/index.jsx [hunk 18]
```diff
@@ -275,14 +279,13 @@ function renderPostDescriptor(
     authorOrNil,
     !hideSubredditLabel && flairOrNil,
   ]);
-  const approvalStatusFlair = renderApprovalStatusFlair(post.approved, post.removed, post.spam);
+  const approvalStatusFlair = renderApprovalStatusFlair(post);
```
Comment (birakattack) [inline src/app/components/Post/PostHeader/index.jsx]: Should be renamed here as well.

File: src/app/components/Post/PostHeader/index.jsx (hunk 17 above)
Comment (birakattack) [inline src/app/components/Post/PostHeader/index.jsx]: Actually this will be fixed with @madbook's PR so we can just ignore this naming change for now as it will happen in subsequent PRs

File: src/app/actions/modTools.js [hunk 19]
```diff
@@ -101,6 +115,27 @@ export const approve = id => async (dispatch, getState) => {
...
 };
 
+export const distinguish = (id, distinguishType) => async (dispatch, getState) => {
+  const state = getState();
+  const apiOptions = apiOptionsFromState(state);
+  const model = modelFromThingId(id, state);
+  const distinguishTypeMap = {
+    'moderator': 'yes',
+    '': 'no',
+  };
```
Comment (madbook) [inline src/app/actions/modTools.js]: 💅 I think that I was confused as to which `distinguish` method I was looking at before – this mapping of 'moderator' to 'yes', etc really should be in the `distinguish` endpoint function in node-api-client.  If another app were to use this endpoint for some reason, they would also need to know to do this mapping before calling `Modtools.distinguish`.  Better to pass in the unmapped value and have Modtools.distinguish handle the mapping internally

**Comment by scarow** [comment]
:haircut_man:

**Comment by madbook** [comment]
🐟
